### ISSUE TITLE
Full installation management via packages.config file

### ISSUE THREAD

**Issue by gregsdennis** [chocolatey/choco#533]
Currently, the manifest file is merely a list of packages for chocolatey to install.  However, it could be used to fully control what is installed on a machine.

For instance, suppose I have a bunch of machines with packages A1.0, B2.2, and C3.0.  I would like to batch alter my machines by:
- Upgrading A to 1.2,
- Uninstalling B completely because I don't need it anymore, and
- Rolling C back to 2.8

Right now, my only option for the uninstall would be a custom script, but the upgrade and (I believe) rollback can be handled by choco.

I would like to send just a manifest file that instructs choco to perform all three of these actions.

I like the manifest file format, so I don't think that should be modified.  (Especially since it's basically Nuget's packages.config.)

Instead of modifying the manifest, the feature could be as simple as adding an install switch like `-image` or `-full` that works with manifest files to indicate that the manifest contains the desired state of all packages installed on the machine.  Choco would then uninstall any packages not listed.

_packages.config_:

```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="a" version="1.2.0"/>
  <package id="c" version="2.8.0"/>
</packages>
```

_command line_:

```
choco install "packages.config" -image
```

**Comment by gregsdennis** [comment]
Of course, an uninstall of an unlisted package would only occur if there are no dependent packages listed.

**Comment by ferventcoder** [comment]
I'm not completely sure about this - I don't want a call to install to automatically go about uninstalling things. There are tools that already provide functionality like this, called configuration managers - like Puppet/Chef. In Puppet it is called purge.

**Comment by ferventcoder** [comment]
Will need to think on this for awhile.

**Comment by gregsdennis** [comment]
The tools out there don't do what we need.  We have a unique tiered deployment strategy that nothing seems to support fully.  With choco we can control that via the manifest files, sending each tier what is supposed to be installed.  If necessary, I can write a shell app that merely calls choco, but integration into choco would be nice.

Secondarily, if you're hung up on the "install" command, it's fine to have a different command.

**Comment by gep13** [comment]
As soon as I read this issue, I immediately thought about Puppet and Chef.

Out of interest, why wouldn't something like Puppet, using the Chocolatey Provider, not fill all your requirements?

**Comment by gregsdennis** [comment]
Hadn't seen puppet before.  Chef is more web deploy whereas we're deploying to many workstations.  Puppet may work, but we'll need to research further.

Even so, I think this would be a nice feature.

**Comment by gregsdennis** [comment]
It looks like Puppet would be a good solution for distribution and configuration management, but we would still need choco for packaging and installation of our apps (we write them; we're not merely distributing 3rd party apps to our machines).  This feature would greatly assist us.

**Comment by gep13** [comment]
>@gregsdennis said...
This feature would greatly assist us.

I am not saying that it wouldn't be a great feature, but I am thinking about whether this should be a core focus of Chocolatey. **NOTE:** I am not saying that it won't be, just playing devil's advocate.

If you look at something like Puppet, using the Chocolatey Provider, as shown in this video:

https://www.youtube.com/watch?v=NNM2H4GsWYA

Puppet already has a mechanism to say that ensure this package is installed, etc.  So I am wondering whether we leave Chocolatey to do it's thing, i.e. provide the mechanism to install, upgrade and uninstall applications, and then use CM tools like Puppet to specify that this is "what" applications should be installed.

Thoughts?

**Comment by gregsdennis** [comment]
@gep13, I see what you're saying.  Looking into puppet (reading & watching videos) I'm not sure that Puppet (or any other available system that I can find) provides the specific feature set that we need for our organization, processes, and architecture.  We're most likely going to have to roll our own management system.  If we need to orchestrate choco as part of that, we're capable.  I'd just prefer it to be built in.

@ferventcoder, I'd like to get your thoughts on our specific design; maybe puppet *can* do what we need.  However, I'd prefer not to discuss it publicly.

**Comment by ferventcoder** [comment]
@gregsdennis Sure - you can shoot me an email - my github username at gmail will do. Or you can send me a message with my first name at puppetlabs dot com.

**Comment by gregsdennis** [comment]
I was pondering this feature a bit more and it seems that choco would be in a better position to manage this kind of feature than an external app because choco has access to the package dependency graph.

If an external app is attempting to do this, I'd have to duplicate the dependencies in that app (be it puppet or something home-grown).  If the external app didn't have this information, it would be forced to simply go down the list in the manifest, which may not be ordered properly.  If it tries to command choco to uninstall a package that is a dependency for another, the uninstall will fail.  This will likely result in a bad state on the machine.

But if choco has this feature, then it can look at the dependency graph, see what needs to be (un)installed, and determine the correct sequence to bring the machine into the desired state.

**Comment by ferventcoder** [comment]
That's a fair assessment.

**Comment by gregsdennis** [comment]
>I don't want a call to install to automatically go about uninstalling things - @ferventcoder 

Does this mean that Chocolatey doesn't behave like Nuget when it comes to dependencies?  For example, in Nuget if I have *A* which depends on *B*, and I upgrade *A* to a version that no longer needs *B*, then *B* will also uninstall (silently) if I have no other packages that need it.

**Comment by ferventcoder** [comment]
> For example, in Nuget if I have A which depends on B, and I upgrade A to a version that no longer needs B, then B will also uninstall (silently) if I have no other packages that need it.

We don't automatically uninstall dependencies. You must explicitly opt-in to that.

~~~

 -x, --forcedependencies, --force-dependencies, --removedependencies, --remove-dependencies
     RemoveDependencies - Uninstall dependencies when uninstalling package(s-
       ). Defaults to false.
~~~

**Comment by gregsdennis** [comment]
That was the interesting thing I saw in Nuget.  During *uninstall* it will default to `false` for uninstalling dependencies, but during *upgrade* it doesn't.  It may just be a Nuget bug/nuance.

**Comment by ferventcoder** [comment]
It's similar with choco.

**Comment by ferventcoder** [comment]
Except that choco doesn't allow use of packages.config for upgrade/uninstall yet.

**Comment by gregsdennis** [comment]
I've been going over this with some of the guys here, trying to implement an external app that does this.  We've determined that *if* a manifest could be used to control the application inventory of a machine, it should have an explicit `uninstall` section.  `install` and `upgrade` could be treated as the same.

**Comment by ferventcoder** [comment]
> It just doesn't seem right to assume that a package should be uninstalled simply because it's not in the manifest.

Sweet. That makes this feel better.

**Comment by ferventcoder** [comment]
Unfortunately this had to bump with lots of other things we wanted in 0.10.9.

**Comment by ferventcoder** [comment]
The bump was due to a required  7z update.

**Comment by ryno24** [comment]
Any ETA on when run choco upgrade package.config will be added? Based on what I can find this has been waiting to go in to a version for over a year, and the workaround of choco install calling beforemodify.ps1 #962 is also waiting on it.

**Comment by jcoutch** [comment]
@ferventcoder / @gep13 - Is there any chance of this getting merged into 0.10.16?

**Comment by Bjego** [comment]
Are there any news on this feature? The uninstallation via packages.config goes hand in hand with my uninstall script posted here: https://github.com/chocolatey/choco/issues/1276

**Comment by TheCakeIsNaOH** [comment]
Here is a wrapper script I use to uninstall a list of packages from a packages.config file.
```
[xml]$packagesXML = Get-Content path\to\packages.config
$packagesXML.packages.package.id | ForEach-Object {
    choco uninstall $_
}
```
Not the most elegant or reliable thing, but it works pretty well if you are able to monitor it.

**Comment by sigma67** [comment]
Is this still on the radar? A bit inconsistent that only `install` supports a manifest.

**Comment by pauby** [comment]
@sigma67 the issue has not been allocated to a milestone so is not being worked on at the moment.

**Comment by TheCakeIsNaOH** [comment]
@sigma67 there is an open community pull request here to add upgrade support: https://github.com/chocolatey/choco/pull/1132

But since the issue is not assigned to milestone of a close version, which is a good indicator that it is not being worked on by the Chocolatey team currently.

**Comment by pauby** [comment]
The functionality requested in the issue is something that a configuration management tool would provide. It can be done by a script. The packages.config file is a list of packages that are to be installed onto a computer with options, a version etc.

This isn't something we would be adding to Chocolatey CLI.

I'm going to go ahead and close this, but we can reopen it if necessary.

**Comment by Masahigo** [comment]
@pauby Hi, what would be then the preferred approach? To me it would seem logical to have the packages.config file sit locally on each VM and have the environment specific package parameters included in the file.

I was hoping to use just a single Chocolatey CLI command to install and/or update Chocolatey packages on target VMs. The **choco upgrade** is idempotent and it would be so nice to use just that instead of having to build extra logic to determine when to use install vs upgrade command against each target VM. Especially when we have to include all the environment specific parameters every time the choco install/upgrade command is executed against target VM. 

The packages.config would be a more clean approach imho.

**Comment by pauby** [comment]
@Masahigo as I mentioned in my [previous comment](https://github.com/chocolatey/choco/issues/533#issuecomment-1924410398), this is the job of a configuration manager.

You are looking to enforce a software state on a computer. That's not the purpose of a [packages.config](https://learn.microsoft.com/en-us/nuget/reference/packages-config) file.

> choco upgrade is idempotent and it would be so nice to use just that instead of having to build extra logic to determine when to use install vs upgrade command against each target VM

By default, `choco upgarde` will install a package if not's installed. You can disable this using `choco feature disable --name=skipPackageUpgradesWhenNotInstalled`.

```
skipPackageUpgradesWhenNotInstalled - Skip Packages Not Installed During Upgrade - if a package is not installed, do not install it during the upgrade process.
```

My suggestion would be to use a configuration manager, or write a script (something like):

```powershell
$packageInstall = @( 'package-a', 'package-b'' )
$packageUninstall = @( 'package-c', 'package-d' )

$packageUninstall | ForEach-Object { choco uninstall -y $_ }
$packageInstall | ForEach-Object { choco upgrade -y $_ }
}
```

**Comment by Masahigo** [comment]
@pauby Thanks for the advice!
